### PR TITLE
feat: basic single-user password protection

### PR DESCRIPTION
**Pull request by keturn**
## What type of PR is this? (check all applicable)

- Feature

## Have you discussed this change with the InvokeAI team?

Consider this a strawman proposal for discussion.
      
## Have you updated all relevant documentation?
- [ ] No

## Description

This allows configuring a password to be required for access to the web server.

The goal is to have a server I can put on an open port without immediately exposing everything to the whole network.

I'm still targeting a single-user use case here — this isn't intended as a step toward building multi-user/multi-tenant hosting in to the basic app.

The implementation in this initial push kinda works, but there are a number of concerns to address before including it as a real feature.

- [ ] should refuse to accept passwords in cleartext, i.e. must require encrypted (HTTPS) transport.
  - _possibly_ with potential to override if that encrypted transport is being provided by some other layer (e.g. proxy, tunnel), but that's debatable. Maybe we say people who are capable of setting up an encrypted proxy have to do their own authorization layer. Most people will be using only this built-in server, and we'd rather not add ways to misconfigure it.
- [ ] I'm using [HTTP Basic Authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#basic_authentication_scheme) here, because I thought that was the simplest thing that could possibly work, but it turns out the UX for Basic Auth is pretty bad in contemporary browsers.
  - My password manager has no access to the username & password fields here in the way that it would for auto-filling HTML forms.
  - The browser doesn't offer a UI to log out. I attempted to add this `/logout` endpoint to encourage the browser to forget its credentials by always sending `401 UNAUTHORIZED`, but that just makes the browser re-prompt again. It's a much more confusing experience than "You are now logged out, goodbye!"
- [ ] I put this in a Middleware on the FastAPI app because I wanted to make sure to cover _any_ and _all_ access to this app without adding an explicit `Depends` to each and every endpoint (which is the way that the [FastAPI Tutorial](https://fastapi.tiangolo.com/tutorial/security/first-steps/#fastapis-oauth2passwordbearer) demonstrates it). But it's left me uncertain if I'm clashing with some guiding FastAPI principle here.
- [ ] There's an open question in my middleware implementation about whether it will fail-open or fail-closed when presented with some unexpected authentication method. https://github.com/keturn/InvokeAI/blob/5b7de495cbca6e216ad3d0c0e991cacde45d9efc/invokeai/app/authn.py#L43-L47

## Related Tickets & Documents

No related tickets found. There have been several instances of people asking on discord about password protection, and there's this discussion: https://github.com/invoke-ai/InvokeAI/discussions/2568

## Added/updated tests?

- [ ] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

No.

### CODE REVIEW

File: invokeai/app/authn.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        # Mostly copy/pasted from the starlette docs. It doesn't ship an implementation of this?
+        auth = conn.headers["Authorization"]
+        try:
+            scheme, credentials = auth.split()
+            if scheme.lower() != "basic":
+                # FIXME: this looks like its designed to fall-through to let another handler deal
+                #   with other schemes, but will that bypass our auth requirements? Or will the lack
+                #   of an authenticated user object prevent the request from getting through?
+                return
```
Comment (keturn) [inline invokeai/app/authn.py:47]: does this fail open or fail closed?

File: invokeai/app/api_app.py [hunk 2]
```diff
@@ -77,7 +77,12 @@
 
 # Create the app
 # TODO: create this all in a method so configuration/etc. can be passed in?
-app = FastAPI(title="Invoke AI", docs_url=None, redoc_url=None, separate_input_output_schemas=False)
+app = FastAPI(
+    title="Invoke AI",
+    docs_url=None,
+    redoc_url=None,
+    separate_input_output_schemas=False,
+)
```
Comment (keturn) [inline invokeai/app/api_app.py:85]: Why are all these whitespace changes in here? Aren't these the _same set_ of changes I _backed out_ of #5262 to get it to pass the format checks then? Why did ruff format add them back this time? :dizzy_face:

**Comment by hipsterusername** [comment]
I think the biggest risk of adding this is maintaining it. Managing auth is not aligned with the core application, and is not something we want to really focus on as a "supported" feature, given it adds complexity to our core worker. 

There may be opportunities to offer more security capabilities in the future, but I'm not convinced we want to deal with auth as a core part of the app

**Comment by keturn** [comment]
> I think the biggest risk of adding this is maintaining it. Managing auth is not aligned with the core application, and is not something we want to really focus on as a "supported" feature, given it adds complexity to our core worker.

Right, which is why I was hoping for a single point of configuration somewhere in the uvicorn/starlette/fastapi stack where we can add it and it applies to the whole thing without the app itself having to be authentication & authorization-aware.

I _think_ adding it as AuthenticationMiddleware has done so, though I should probably confirm that websockets don't somehow run a side channel around it or something like that.

**Comment by hipsterusername** [comment]
I'm very wary of adding anything like this, because we aren't intending to maintain or support it long-term.

**Comment by keturn** [comment]
In my case, these are things that it's nice to have even for usage on LAN.

- When debugging latency and caching issues, fewer layers of proxy/tunnel/etc seems better, so HTTPS is better than HTTP-over-SSH-tunnel. Also avoids the problem where the chromebook's SSH client would shut down when the InvokeAI window started overloading the memory pressure.
- 21st century web protocols (HTTP/2, and the QUIC-based HTTP/3) _require_ TLS. I haven't run the benchmarks to back it up, but I think there's _potential_ for improved performance there on certain types of connections or scenarios when the application generates a lot of requests.
- Just because my LAN isn't _as dangerous_ as the Internet, it doesn't mean I want every person and IoT device on the network to have free access to this app.

Overall, I think if you don't want an application to be on the network, that is a _strong indication_ that it should never start a network service in the first place.

> My position is: If you want auth, maintain a wrapper/fork yourself.

That's not an unreasonable position, but the current way `invokeai-web` and `api_app` start the server is not exa-- actually, hold on, looking over it again the situation is better than my first impression suggested.
- `socket` is only imported for the case where it needs to find its own port number.
- `asyncio` is imported to create an event loop, but not one that's passed through to any invokeai Services.
- `SocketIO` I am leery of. It creates this thing on module import [here](https://github.com/invoke-ai/InvokeAI/blob/f0c70fe3f19e59d2be746e00a4e16029569562dd/invokeai/app/api_app.py#L88) but that `socket_io` object is never referenced again, and -- wait, am I just being mislead by the terminology here? Does the "Server" object of a "Socket.IO" package not actually open any network sockets for listening on the server???

**Comment by keturn** [comment]
I finally got around to putting this in an independent package: https://gitlab.com/keturn/invokeai-authn

**Pull request closed without merging**